src/smt.cpp

```diff
@@ -173,19 +197,102 @@ string or_omit(const std::vector<expr> &evec) {
   return s;
 }
 
+Expr::Expr(std::optional<z3::expr> &&z3_expr) {
+  this->z3_expr = std::move(z3_expr);
+}
+
+Expr Expr::simplify() const {
+  auto z3_expr = fmap(this->z3_expr, [](auto e) { return e.simplify(); });
+
+  return Expr(std::move(z3_expr));
+}
 
-Expr Expr::urem(const Expr& rhs) const {
-  return {fmap(z3_expr, [&](auto e) { return z3::urem(e, *rhs.z3_expr); })};
+ExprVec Expr::toNDIndices(const ExprVec &dims) const {
+  assert(dims.size() > 0);
+
+  auto idx_1d = *this;
+  ExprVec expanded_exprs;
+  expanded_exprs.reserve(dims.size());
+  std::for_each(dims.crbegin(), dims.crend(), 
+    [&idx_1d, &expanded_exprs](const Expr& dim) {
+      expanded_exprs.push_back(idx_1d.urem(dim));
+      idx_1d = idx_1d.udiv(dim);
+    });
+  std::reverse(expanded_exprs.begin(), expanded_exprs.end());
+  return expanded_exprs;
+}
+
+Expr Expr::urem(const Expr &rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::urem(e, *rhs.z3_expr); });
+  
+  return Expr(std::move(z3_expr));
 }
 
 Expr Expr::udiv(const Expr& rhs) const {
-  return {fmap(z3_expr, [&](auto e) { return z3::udiv(e, *rhs.z3_expr); })};
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::udiv(e, *rhs.z3_expr); });
+  
+  return Expr(std::move(z3_expr));
 }
 
-Expr Expr::simplify() const {
-  return {fmap(z3_expr, [](auto e) { return e.simplify(); })};
+Expr Expr::add(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e + *rhs.z3_expr; });
+  
+  return Expr(std::move(z3_expr));
 }
 
+Expr Expr::sub(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e - *rhs.z3_expr; });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::mul(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e * *rhs.z3_expr; });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::ult(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::ult(e, *rhs.z3_expr); });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::ugt(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::ugt(e, *rhs.z3_expr); });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::boolAnd(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e && *rhs.z3_expr; });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::boolOr(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e || *rhs.z3_expr; });
+  
+  return Expr(std::move(z3_expr));
+}
+
+Expr Expr::bvVal(const uint32_t val, const size_t sz) {
+  auto z3_expr = sctx.z3Map([val, sz](auto &ctx){ return ctx.bv_val(val, sz); });
```

Comment:
Would fmap work?

Reply:
That also works, if we define another fmap.
```c++
template <class T, class Fn>
auto fmap(optional<T> &x, Fn &&fn) {
...
}
```

Note that this one does not have **const** in the parameter type.

Reply:
Implemented as `fupdate()`